dandi/download.py

```diff
@@ -501,3 +503,66 @@ def _download_file(
         os.utime(path, (time.time(), ensure_datetime(mtime).timestamp()))
 
     yield {"status": "done"}
+
+
+class DownloadDirectory:
+    def __init__(self, filepath, filehash):
+        #: The path to which to save the file after downloading
+        self.filepath = Path(filepath)
+        #: The expected hash of the downloaded data
+        self.filehash = filehash
+        #: The working directory in which downloaded data will be temporarily
+        #: stored
+        self.dirpath = append_ext(self.filepath, ".dandidownload")
+        #: The file in `dirpath` to which data will be written as it is
+        #: received
+        self.writefile = self.dirpath / self.filepath.name
+        #: A `fasteners.InterProcessLock` on `dirpath`
+        self.lock = None
+        #: An open filehandle to `writefile`
+        self.fp = None
+        #: How much of the data has been downloaded so far
+        self.offset = None
+
+    def __enter__(self):
+        from fasteners import InterProcessLock
+
+        self.dirpath.mkdir(parents=True, exist_ok=True)
+        self.lock = InterProcessLock(str(append_ext(self.writefile, ".lock")))
+        if not self.lock.acquire(blocking=False):
+            raise RuntimeError("Could not acquire download lock for {self.filepath}")
+        chkpath = append_ext(self.writefile, ".checksum")
+        try:
+            checksum = chkpath.read_text(encoding="utf-8").strip()
+        except FileNotFoundError:
+            checksum = ""
+        if checksum == self.filehash:
+            # Pick up where we left off, writing to the end of the file
+            self.fp = self.writefile.open("ab")
+        else:
+            # Truncate the file (if it even exists)
+            self.fp = self.writefile.open("wb")
```

Comment:
could you please explicitly `unlink` old one first?  I juts fear that if somehow this file was saved into git-annex, it would be read-only, and here it would just fail.  Sure thing the same could happen also for `"ab"` mode, but then it would require annex specific (`git annex unlock` or `datalad unlock`) and we better not introduce such here ATM.  But since workaround for the other case is easy (unlink first) I think we should just do it.

Reply:
Done.